libs/ballot-interpreter-nh/src/convert.test.ts

```diff
@@ -62,8 +62,8 @@ test('converting the Hudson ballot', async () => {
     issues: [],
   });
 
-  expect(JSON.stringify(convertResult.election, null, 2)).toEqual(
-    electionGridLayoutNewHampshireHudsonFixtures.electionDefinition.electionData
+  expect(convertResult.election).toMatchObject(
+    electionGridLayoutNewHampshireHudsonFixtures.electionDefinition.election
```

Comment:
Some testing issues were also due to spacing differences, which can be resolved by not comparing strings but rather comparing objects.

Reply:
There is a comment above this:

```ts  
  // uncomment this to update the fixture
  // require('fs').writeFileSync(
  //   require('path').join(
  //     __dirname,
  //     '../../fixtures/data/electionGridLayoutNewHampshireHudson/election.json'
  //   ),
  //   JSON.stringify(convertResult.election, null, 2),
  //   'utf8'
  // );
```

I think it is simply assuming that you update the fixture as such, when I actually updated the fixture manually (so `prettier` applied).

Serialization does affect election hashes, though, because when we parse election definitions we're not stripping out whitespace or new lines.

Reply:
@eventualbuddha is there a reason we don't strip out spacing when we get the `electionData` or when we compute the `electionHash`?

Reply:
I think this gets back to the canonicalization problem that @benadida talks about. The claim is that it's hard to canonicalize JSON (e.g. by stripping whitespace) in a way that is consistent across implementations. On the other hand, it's relatively easy to compute the same SHA256 hash from the same UTF-8 bytes. I think this is another instance of wanting to ensure that we always produce the same data with the same hash given the same input.

Reply:
It's more than convenience, there are so many bugs that can occur when you're relying on authenticity of a data structure that assumes canonicalization. Consider things like duplicate keys in a JSON string, does that get canonicalized in the same way as it is interpreted by the app logic?

Really important to always have the original bytes of any file that requires authentication (which is going to be all the inputs and outputs.)

Reply:
OK, thanks for the background! This explains to me why we don't try to canonicalize within the app. Rather, we deal in the original bytes of the file.

For the two tests being flagged here, however, I'm still under the impression that I don't need to test on the exact formatting of the outputted election definitions, as strings. One test is in `ballot-interpreter-nh`'s `convert` module and one is in `converter-ms-sems`. These modules output an election definition in bytes, and then the application and our system use it from then onward as unchanging bytes. What does matter is that the app does not try to canonicalize, it just takes the file as is. The exact spacing and formatting of that output file does not matter as long as it can be parsed. Am I missing something?